### PR TITLE
Feat/miningpool mw

### PR DESCRIPTION
**Pull request by Geapefurit**

### CODE REVIEW

File: npool/miningpool/gw/v1/app/pool/pool.proto [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+import "google/api/annotations.proto";
+import "npool/basetypes/miningpool/v1/enums.proto";
+
+// Service Name
+service Gateway {
+  rpc AdminCreatePool(AdminCreatePoolRequest)
+      returns (AdminCreatePoolResponse) {
+    option (google.api.http) = {
+      post : "/v1/admin/create/apppool"
```
Comment (kikakkz) [inline npool/miningpool/gw/v1/app/pool/pool.proto]: appminingpool

File: npool/miningpool/gw/v1/app/pool/pool.proto [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  string Logo = 71;
+  string Description = 80;
+  repeated Coin Coins = 90;
+  repeated FractionRule FractionRules = 100;
+}
+
+message Coin {
+  string EntID = 10;
+  string PoolID = 20;
+  basetypes.miningpool.v1.CoinType CoinType = 30;
```
Comment (kikakkz) [inline npool/miningpool/gw/v1/app/pool/pool.proto]: Unify coin usage with whole system

File: npool/miningpool/gw/v1/app/pool/pool.proto [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  bool FixedRevenueAble = 60;
+  string LeastTransferAmount = 70;
+  uint32 BenefitIntervalSeconds = 80;
+  string Remark = 90;
+  basetypes.miningpool.v1.MiningpoolType MiningpoolType = 100;
+}
+
+message FractionRule {
+  string EntID = 10;
+  string CoinID = 20;
```
Comment (kikakkz) [inline npool/miningpool/gw/v1/app/pool/pool.proto]: Unify coin usage with whole system

File: npool/miningpool/gw/v1/fraction/fraction.proto [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+message Fraction {
+  uint32 ID = 10;
+  string EntID = 20;
+  string AppID = 30;
+  string UserID = 40;
+  string OrderUserID = 50;
+  basetypes.miningpool.v1.WithdrawState WithdrawState = 60;
+  uint32 WithdrawTime = 70;
```
Comment (kikakkz) [inline npool/miningpool/gw/v1/fraction/fraction.proto]: WithdrawAt

File: npool/miningpool/gw/v1/fraction/fraction.proto [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  uint32 PayTime = 80;
```
Comment (kikakkz) [inline npool/miningpool/gw/v1/fraction/fraction.proto]: PaidAt

File: npool/miningpool/gw/v1/fraction/fraction.proto [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  string Msg = 90;
+  uint32 CreatedAt = 100;
+  uint32 UpdatedAt = 110;
+}
+
+message CreateFractionRequest {
+  string AppID = 10;
+  string UserID = 20;
+  string OrderUserID = 30;
```
Comment (kikakkz) [inline npool/miningpool/gw/v1/fraction/fraction.proto:50]: I think we should use OrderID here, or this API should be implemented by order-gateway

File: npool/miningpool/gw/v1/gooduser/gooduser.proto [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  string CoinID = 90;
+  basetypes.miningpool.v1.CoinType CoinType = 100;
+  basetypes.miningpool.v1.RevenueType RevenueType = 110;
+  string FeeRatio = 120;
+  uint32 CreatedAt = 130;
+  uint32 UpdatedAt = 140;
+}
+
+message AdminCreateGoodUserRequest {
+  string CoinID = 10;
```
Comment (kikakkz) [inline npool/miningpool/gw/v1/gooduser/gooduser.proto]: Use GoodID here. Try to get information with GoodID.

File: npool/miningpool/gw/v1/gooduser/gooduser.proto [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+  string RootUserID = 20;
+  float HashRate = 30;
```
Comment (kikakkz) [inline npool/miningpool/gw/v1/gooduser/gooduser.proto]: 1  Float is not safe.
2  Get good units from good information

File: npool/miningpool/gw/v1/gooduser/gooduser.proto [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+import "google/api/annotations.proto";
+import "npool/basetypes/miningpool/v1/enums.proto";
+
+// Service Name
+service Gateway {
+  rpc AdminCreateGoodUser(AdminCreateGoodUserRequest)
+      returns (AdminCreateGoodUserResponse) {
+    option (google.api.http) = {
+      post : "/v1/admin/create/gooduser"
```
Comment (kikakkz) [inline npool/miningpool/gw/v1/gooduser/gooduser.proto]: goodminingpooluser

File: npool/miningpool/gw/v1/fraction/fraction.proto [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+option go_package = "github.com/NpoolPlatform/message/npool/miningpool/gw/v1/fraction";
+
+import "google/api/annotations.proto";
+import "npool/basetypes/miningpool/v1/enums.proto";
+
+// Service Name
+service Gateway {
+  rpc CreateFraction(CreateFractionRequest) returns (CreateFractionResponse) {
+    option (google.api.http) = {
+      post : "/v1/create/fraction"
```
Comment (kikakkz) [inline npool/miningpool/gw/v1/fraction/fraction.proto]: miningpoolfrationwithdraw

File: npool/miningpool/gw/v1/orderuser/orderuser.proto [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+option go_package = "github.com/NpoolPlatform/message/npool/miningpool/gw/v1/orderuser";
+
+import "google/api/annotations.proto";
+import "npool/basetypes/miningpool/v1/enums.proto";
+
+// Service Name
+service Gateway {
+  rpc GetOrderUser(GetOrderUserRequest) returns (GetOrderUserResponse) {
+    option (google.api.http) = {
+      post : "/v1/get/orderuser"
```
Comment (kikakkz) [inline npool/miningpool/gw/v1/orderuser/orderuser.proto]: orderminingpooluser

File: npool/miningpool/gw/v1/pool/pool.proto [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+option go_package = "github.com/NpoolPlatform/message/npool/miningpool/gw/v1/pool";
+
+import "google/api/annotations.proto";
+import "npool/basetypes/miningpool/v1/enums.proto";
+
+// Service Name
+service Gateway {
+  rpc AdminGetPools(AdminGetPoolsRequest) returns (AdminGetPoolsResponse) {
+    option (google.api.http) = {
+      post : "/v1/admin/get/pools"
```
Comment (kikakkz) [inline npool/miningpool/gw/v1/pool/pool.proto]: miningpools

File: npool/miningpool/gw/v1/rootuser/rootuser.proto [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+
+import "google/api/annotations.proto";
+import "npool/basetypes/miningpool/v1/enums.proto";
+
+// Service Name
+service Gateway {
+  rpc AdminCreateRootUser(AdminCreateRootUserRequest)
+      returns (AdminCreateRootUserResponse) {
+    option (google.api.http) = {
+      post : "/v1/admin/create/rootuser"
```
Comment (kikakkz) [inline npool/miningpool/gw/v1/rootuser/rootuser.proto]: roogminingpooluser

File: npool/miningpool/mw/v1/gooduser/gooduser.proto [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+  rpc ExistGoodUserConds(ExistGoodUserCondsRequest)
+      returns (ExistGoodUserCondsResponse) {}
+  rpc UpdateGoodUser(UpdateGoodUserRequest) returns (UpdateGoodUserResponse) {}
+  rpc DeleteGoodUser(DeleteGoodUserRequest) returns (DeleteGoodUserResponse) {}
+}
+
+message GoodUserReq {
+  optional uint32 ID = 10;
+  optional string EntID = 20;
+  optional string CoinID = 30;
```
Comment (kikakkz) [inline npool/miningpool/mw/v1/gooduser/gooduser.proto]: Unify CoinID to CoinTypeID

**Review by kikakkz — CHANGES_REQUESTED** [review]

File: npool/miningpool/gw/v1/app/pool/pool.proto (hunk 3 above)
Comment (kikakkz) [reply npool/miningpool/gw/v1/app/pool/pool.proto:76]: PoolCoinTypeID

**Review by kikakkz — CHANGES_REQUESTED** [review]

File: npool/miningpool/gw/v1/fraction/fraction.proto (hunk 6 above)
Comment (Geapefurit) [reply npool/miningpool/gw/v1/fraction/fraction.proto:50]: should be implemented by order-gateway

**Review by kikakkz — APPROVED** [review]

**Pull request merged**